clippy_lints/src/matches/manual_utils.rs

```diff
@@ -274,3 +274,71 @@ pub(super) fn try_parse_pattern<'tcx>(
 fn is_none_expr(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
     is_res_lang_ctor(cx, path_res(cx, peel_blocks(expr)), OptionNone)
 }
+
+fn expr_ty_adjusted(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+    cx.typeck_results()
+        .expr_adjustments(expr)
+        .iter()
+        // We do not care about exprs with `NeverToAny` adjustments, such as `panic!` call.
+        .any(|adj| !matches!(adj.kind, Adjust::NeverToAny))
+}
+
+fn expr_has_type_coercion<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'tcx>) -> bool {
+    if expr.span.from_expansion() {
+        return false;
+    }
```

Comment:
hmmm, I forgot why I left it there, but... you're right 😂

Reply:
Oh I think I got it, it was a lazy solution to bypass this test:

```rust
match Some(0) {
    Some(x) => Some(vec![x]),
     None => None,
};
// ^ help: try: `Some(0).map(|x| vec![x])`
```

Reply:
Ok, so... I think where the problem was.

technically, the above case is not a coercion site(?) according to the coercion doc:
> ...these are typically places where the desired type is explicit or can be derived by propagation from explicit types (without type inference).

Thus using `map` here actually works. However, because I have a type adjustment check at first, which causing it to return early as soon as any expr's ty was adjusted (and [this expr in `vec!`](https://doc.rust-lang.org/1.81.0/src/alloc/macros.rs.html#54) was indeed adjusted), causing FN on the above test case.

So, ignore what I said earlier that it was ~totally irrelevant~, here is a counter example which might not seems rare at all:

```rust
let x: Option<Vec<&[u8]>> = match Some(()) {
    Some(_) => Some(vec![b"1234"]),
    None => None,
};
```

With that `from_expansion` check, this will surely cause FP. But I also needed that type adjustment check, otherwise this counter example wouldn't work.

Now I'm really confused, do I need to bring back the `expr_has_explicit_ty` function from the beginning?